apollo-federation/src/query_plan/query_planner.rs

```diff
@@ -325,14 +325,18 @@ impl QueryPlanner {
     ) -> Result<QueryPlan, FederationError> {
         let operation = document
             .operations
-            .get(operation_name.as_ref().map(|name| name.as_str()))?;
-
+            .get(operation_name.as_ref().map(|name| name.as_str()))
+            .map_err(|_| {
+                if operation_name.is_some() {
+                    SingleFederationError::UnknownOperation
+                } else {
+                    SingleFederationError::OperationNameNotProvided
+                }
+            })?;
         if operation.selection_set.is_empty() {
             // This should never happen because `operation` comes from a known-valid document.
             // We could panic here but we are returning a `Result` already anyways, so shrug!
```

Comment:
```suggestion
```

Let's remove this line (`bail!` technically doesn't panic, but it should be used as if it does)